### PR TITLE
feat(ts): convert menu

### PR DESCRIPTION
**Pull request by Haroenv**
**Summary**

converts the menu-related files to typescript

**Result**

converted:
- connectMenu
- menu
- menuSelect

fixed a bug in connectMenu#getRenderState (forgotten to nest by attribute)

### CODE REVIEW

File: src/connectors/menu/connectMenu.ts [hunk 1]
```diff
@@ -115,30 +139,33 @@ export default function connectMenu(renderFn, unmountFn = noop) {
...
+    let _createURL: MenuRendererOptions['createURL'] | undefined;
+    let _refine: MenuRendererOptions['refine'] | undefined;
 
     // Provide the same function to the `renderFn` so that way the user
     // has to only bind it once when `isFirstRendering` for instance
+    let isShowingMore = false;
     let toggleShowMore = () => {};
+    function createToggleShowMore(
+      renderOptions: RenderOptions,
+      widget: Widget
```
Comment (Haroenv) [inline src/connectors/menu/connectMenu.ts:152]: passing the "this" to this function is the most natural method I found for show more. refinementList and hierarchicalMenu aren't yet converted, so this is the first widget with this pattern

File: src/connectors/menu/connectMenu.ts [hunk 2]
```diff
@@ -175,67 +202,74 @@ export default function connectMenu(renderFn, unmountFn = noop) {
       getRenderState(renderState, renderOptions) {
         return {
           ...renderState,
-          menu: this.getWidgetRenderState(renderOptions),
+          menu: {
+            ...renderState.menu,
+            [attribute]: this.getWidgetRenderState(renderOptions),
```
Comment (Haroenv) [inline src/connectors/menu/connectMenu.ts:207]: note that this fixes a bug!

File: src/lib/utils/prepareTemplateProps.ts [hunk 3]
```diff
@@ -4,59 +4,65 @@ import { Template } from '../../types';
...
-  templates: Templates;
-  useCustomCompileOptions: { [key: string]: boolean };
-};
-
-function prepareTemplates(
-  defaultTemplates: Templates = {},
-  templates: Templates = {}
+function prepareTemplates<TTemplates extends Templates>(
+  // can not use = {} here, since the template could have different constraints
+  defaultTemplates?: TTemplates,
```
Comment (Haroenv) [inline src/lib/utils/prepareTemplateProps.ts:20]: not super happy with this, but refinementList for example doesn't pass defaultTemplates at all

File: src/types/widget.ts [hunk 4]
```diff
@@ -217,6 +221,12 @@ export type IndexRenderState = Partial<{
     CurrentRefinementsRendererOptions,
     CurrentRefinementsConnectorParams
   >;
+  menu: {
+    [attribute: string]: WidgetRenderState<
+      MenuRendererOptions,
+      MenuConnectorParams
+    >;
+  };
```
Comment (Haroenv) [inline src/types/widget.ts:229]: same bugfix, as well as noting its omission in the type

File: src/widgets/menu-select/__tests__/menu-select-test.ts [hunk 5]
```diff
@@ -1,19 +1,26 @@
...
   return module;
 });
+const render = _originalRender as jest.MockedFunction<typeof _originalRender>;
 
 describe('menuSelect', () => {
   describe('Usage', () => {
     it('throws without container ', () => {
       expect(() => {
+        // @ts-ignore
         menuSelect({ container: undefined });
```
Comment (Haroenv) [inline src/widgets/menu-select/__tests__/menu-select-test.ts:24]: maybe we should bump the ts version so we can use ts-expect-error?

File: src/widgets/menu/__tests__/__snapshots__/menu-test.ts.snap [hunk 6]
```diff
@@ -17,20 +17,7 @@ Object {
...
-      "value": undefined,
-    },
-    Object {
-      "label": "bar",
-      "transformed": true,
-      "url": "#",
-      "value": undefined,
-    },
-  ],
+  "facetValues": Array [],
```
Comment (Haroenv) [inline src/widgets/menu/__tests__/__snapshots__/menu-test.ts.snap:20]: todo: find which change in the tests caused this to change (likely something using the ts-safe result constructors

File: src/connectors/menu/__tests__/connectMenu-test.ts [hunk 7]
```diff
@@ -499,24 +507,23 @@ See documentation: https://www.algolia.com/doc/api-reference/widgets/menu/js/#co
             results: new SearchResults(helper.state, [
               createSingleSearchResponse({
                 hits: [],
-                facets: {
-                  brand: 300,
-                },
```
Comment (Haroenv) [inline src/connectors/menu/__tests__/connectMenu-test.ts:504]: why is this removed?

File: src/connectors/menu/__tests__/connectMenu-test.ts [hunk 8]
```diff
@@ -533,11 +540,10 @@ See documentation: https://www.algolia.com/doc/api-reference/widgets/menu/js/#co
       const helper = jsHelper(
         createSearchClient(),
         'indexName',
-        menu.getWidgetSearchParameters(new SearchParameters(), { uiState: {} })
+        menu.getWidgetSearchParameters!(new SearchParameters(), { uiState: {} })
       );
 
       const renderState1 = menu.getWidgetRenderState(
-        {},
         createInitOptions({ helper })
```
Comment (Haroenv) [inline src/connectors/menu/__tests__/connectMenu-test.ts:547]: this test was wrong :o

**Review by Haroenv — COMMENTED** [review]
https://github.com/algolia/instantsearch.js/pull/4652/files#diff-3cdf1a39cd7976ea570f612bfa90b1b182c136eab976ec2b4fb528d6b1f27350R101 requires https://github.com/algolia/algoliasearch-helper-js/pull/811

File: src/connectors/menu/connectMenu.ts [hunk 9]
```diff
@@ -175,67 +202,74 @@ export default function connectMenu(renderFn, unmountFn = noop) {
...
             helper,
             attribute,
-            widgetType: this.$$type,
+            widgetType: this.$$type!,
           });
         }
 
-        if (!this._createURL) {
-          this._createURL = facetValue =>
-            createURL(helper.state.toggleRefinement(attribute, facetValue));
```
Comment (Haroenv) [inline src/connectors/menu/connectMenu.ts:202]: this function is deprecated, toggleFacetRefinement its exact replacement

File: src/connectors/menu/__tests__/connectMenu-test.ts (hunk 7 above)
Comment (Haroenv) [reply src/connectors/menu/__tests__/connectMenu-test.ts:504]: because it needs to be nested { brand: {Apple: 300}}, but this test doesn't care about that

**Review by yannickcr — APPROVED** [review]

File: src/connectors/menu/__tests__/connectMenu-test.ts (hunk 8 above)
Comment (eunjae-lee) [reply src/connectors/menu/__tests__/connectMenu-test.ts:547]: good job by TS finding it out :)

**Review by eunjae-lee — APPROVED** [review]
It looks good to me. Congrats on one less widget in JS :)
